#696 create risk log visuals

z-roth opened the pull request:
## Changes

- Added UI components for risk log feature (display risks, delete completed risk, create change request for risk, add new risk)

## Notes

- One of the tests that checks that the Convert to CR button renders doesn't work for some strange reason, but commented it out
- Doesn't like to work with dark mode, not sure how to implement that yet

## Test Cases

### CheckList
- Renders pageblock with given title and header right 
- Renders correct buttons for both resolved and unresolved risks (delete, convert to CR)
- Checkboxes that are marked as checked in the data are checked

### RiskLog
- Renders the Risk Log title
- Renders all risks

## Screenshots (if applicable)

![image](https://user-images.githubusercontent.com/80188262/180303794-2f6c7a5e-823d-41c5-8359-486d463f8f1f.png)

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/docs/ContributorGuide.md) and reach out to your squad if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors
- [x] No newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (if applicable)
- [x] Remove any not-applicable sections
- [x] Assign the PR to yourself
- [x] PR is linked to the ticket
- [x] No `package-lock.json` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack

Closes #696

anthonybernardi commented on src/frontend/pages/ProjectDetailPage/project-view-container/project-view-container.tsx:
```diff
@@ -15,9 +15,11 @@ import PageTitle from '../../../layouts/page-title/page-title';
 import './project-view-container.module.css';
 import { routes } from '../../../../shared/routes';
 import ProjectGantt from './project-gantt/project-gantt';
+import { truncateSync } from 'fs';
```
pls delete

anthonybernardi commented:
The code looks good but I'll do a full code review when everything is more finalized. As far as design goes, I'd prefer:
1. each risk row to be thinner. The whole thing is very large rn
2. i also might want a way to make adding them harder. not sure what the best design for that would be
3. for dark mode, you're gonna have to mess with the listgroupitem or something idk

jamescd18 commented:
Making it harder to create could be as simple as putting it in a modal. 

Also you might wanna do away with dividing lines and background on each Risk. Different bootstrap component (or just a divs) might help with this. 

Also if you wanna get fancier with UI, you make the right side buttons be icons instead, only show on hover, and maybe add tooltips to say their function after mouse hovers for 2 seconds. This is very justifiable because rn as buttons, you're calling more attention to them than they really deserve. The number of primary buttons on a page at a given time should be kept meaningfully minimal.

z-roth commented:
Changed around a few things:
![image](https://user-images.githubusercontent.com/80188262/181009479-6a69a0f7-c883-4ea9-bd75-73e2add8c930.png)
![image](https://user-images.githubusercontent.com/80188262/181009759-1b3adca4-b9d4-46d6-b4ca-19b762580e4e.png)

- Removed ListGroup and ListGroupItem and replaced with divs for cleaner look, also works fine now with dark mode
- Removed input text box and added pop-up modal that prompts user for risk input

Going to look into how to add some icons and tooltips.

z-roth commented:
@jamescd18 are there any pre-existing examples of the use of icons and tooltips that I can look at to implement those UI improvements?

RChandler234 commented:
> @jamescd18 are there any pre-existing examples of the use of icons and tooltips that I can look at to implement those UI improvements?

There should be an example of tooltips on the implemented changes section of the change request detail page. Not sure about making the button an icon, I'd look up a react-bootstrap example or smthn

RChandler234 commented on src/frontend/components/check-list/check-list.tsx:
```diff
@@ -0,0 +1,98 @@
+interface CheckListProps {
+  title: string;
+  headerRight?: ReactNode;
+  list: {
```
Could we have a more descriptive name for this prop, especially because it's asking for these custom jsx objects. I'm thinking something like listItems?

RChandler234 commented on src/frontend/components/check-list/check-list.tsx:
```diff
@@ -0,0 +1,98 @@
+
+interface CheckListProps {
+  title: string;
+  headerRight?: ReactNode;
```
What is this prop for? I'm not seeing it used in the risk log file?

z-roth replied:
I tried to model this component to be modeled after the BulletList component, which also had it, so I figured I would add it in here too (I'm assuming you mean the headerRight prop?)

anthonybernardi replied:
yeah lets keep it for the same reason bulletlist has it, just in case

RChandler234 commented on src/frontend/components/check-list/check-list.tsx:
```diff
@@ -0,0 +1,98 @@
+        </Button>
+      </Modal.Footer>
+    </Modal>
+  );
```
This modal should arguably be in its own file, especially given that we're gonna have to add in the functionality to actually add and save the risk in the database later.

Separate question @jamescd18 @anthonybernardi do you think it'd be worth making a component for modals like this? Seems like it could be something we might end up using a lot.

anthonybernardi replied:
not a bad idea, but will depend on how similar they all are

RChandler234 commented on src/frontend/components/check-list/check-list.tsx:
```diff
@@ -0,0 +1,98 @@
+    setChecks(updatedChecks);
+  };
+
+  const builtList = checks.map((check, idx) => (
+    <div key={idx} className={styles.container}>
+      <Form.Check
+        label={
+          <p
+            style={check.resolved ? { textDecoration: 'line-through' } : { textDecoration: 'none' }}
+          >
+            {check.details}
+          </p>
+        }
+        defaultChecked={check.resolved}
+        data-testId={`testCheckbox${idx}`}
+        onChange={() => handleCheck(idx)}
+      />
+      {check.resolved ? (
+        <Button variant="danger">Delete</Button>
+      ) : (
+        <Button variant="success">Convert to CR</Button>
+      )}
+    </div>
+  ));
+
+  builtList.push(
+    <div>
```
So for this whole section it looks like you're taking in checks that you've declared elsewhere and then using it to render all these react components. I would reformat this so that bulletlist is kinda a custom bit of React in this file. I believe
`const BulletList = (checkList: any) => { return ()}` should work and then you put all the code for the map inside a div inside the return statement (maybe wrap that in a div) and I think that'll work and then you should be able to just add the Add New Risk button below it in the top div rather than appending it to the list of jsx components as you've done here. 

I also suggest this method because it allows you to call bulletlist below like `<BulletList checkList = {checks} />` rather than `{builtList}`; I think this looks a lil cleaner/ makes it more clear what bulletlist is/ what it takes in.

Let me know if this suggestion is confusing btw, I know it's a lot. It's a lil easier to show in code, so I can hop in a call or smthn if need be.

RChandler234 commented on src/frontend/pages/ProjectDetailPage/project-view-container/risk-log/risk-log.tsx:
```diff
@@ -0,0 +1,16 @@
+import CheckList from '../../../../components/check-list/check-list';
+
+interface RiskLogProps {
+  risks: { details: string; resolved: boolean }[];
```
It might be worth adding this risk type to the types folder in utils as it's probably going to be used in several places once we actually hook up the feature. Thoughts @jamescd18?

Also the feature looks great btw! All my comments are just knitpicky stuff.

anthonybernardi replied:
definitely agree with this, there actually should be one already

anthonybernardi commented:
i'm good with this from a visual standpoint. I'll leave it to Reid for the nitty gritty of the code

RChandler234 commented:
This looks a lot cleaner! There might be some stuff we wanna change/ add to the checklist component later, but for the purpose of this ticket this is awesome. Gonna go ahead and approve.

anthonybernardi reviewed (approved):
dub

z-roth commented:
lgtm !!
